src/neo/IO/Caching/HashSetCache.cs

```diff
@@ -0,0 +1,91 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Neo.IO.Caching
+{
+    public class HashSetCache<T> : IReadOnlyCollection<T> where T : IEquatable<T>
+    {
+        /// <summary>
+        /// Sets where the Hashes are stored
+        /// </summary>      
+        private readonly LinkedList<HashSet<T>> sets = new LinkedList<HashSet<T>>();
+
+        /// <summary>
+        /// Maximum capacity of each bucket inside each HashSet of <see cref="sets"/>.
+        /// </summary>        
+        private readonly int bucketCapacity;
+
+        /// <summary>
+        /// Maximum number of buckets for the LinkedList, meaning its maximum cardinality.
+        /// </summary>
+        private readonly int maxBucketCount;
+
+        /// <summary>
+        /// Entry count
+        /// </summary>
+        public int Count { get; }
+        
+        public HashSetCache(int bucketCapacity, int maxBucketCount = 10)
+        {
+            if (bucketCapacity <= 0) throw new ArgumentOutOfRangeException($"{nameof(bucketCapacity)} should be greater than 0");
+            if (maxBucketCount <= 0) throw new ArgumentOutOfRangeException($"{nameof(maxBucketCount)} should be greater than 0");
+
+            this.Count = 0;
+            this.bucketCapacity = bucketCapacity;
+            this.maxBucketCount = maxBucketCount;
+            sets.AddFirst(new HashSet<T>());
+        }
+
+        public bool Add(T item)
+        {
+            if (Contains(item)) return false;
+            Count++;
+            if (sets.First.Value.Count < bucketCapacity) return sets.First.Value.Add(item);
+            var newSet = new HashSet<T>
+            {
+                item
+            };
+            sets.AddFirst(newSet);
+            if (sets.Count > maxBucketCount) sets.RemoveLast();
+            return true;
+        }
+
+        public bool Contains(T item)
+        {
+            foreach (var set in sets)
```

Comment:
We are inserting the items in the last set, but I think that usually we will have usually the most queried items in the last entries. It's an intuition, will be great if we can ensure that.

Reply:
Agree, @shargon. Good idea.

Reply:
in this sense, better to run a normal `for` loop and start from the last index.

Reply:
Usually maxBucketCount is not to many, default is 10, so traversal is very fast, cost about 1E-06s. Whether it is forward traversal or reverse traversal, the speed is almost the same, so the optimization effect is very limited. PS: We are inserting the items in the first set not the last 😂

Reply:
> We are inserting the items in the first set not the last 😂

hahah you are right